File: lib/engine/game_engine/game_state_factory.dart
import 'package:ludo_arena/models/enums.dart';
import 'package:ludo_arena/models/game_config.dart';
import 'package:ludo_arena/models/game_state.dart';
import 'package:ludo_arena/models/player.dart';
import 'package:ludo_arena/models/token.dart';

/// Builds a fresh offline GameState from [GameConfig].
class GameStateFactory {
  static GameState create({
    required GameConfig config,
    String id = 'local',
    DateTime? now,
  }) {
    final seats = PlayerSeat.values.take(config.playerCount).toList();
    final slots = config.slots.isNotEmpty
        ? config.slots
        : [
            for (var i = 0; i < seats.length; i++)
              PlayerSlotConfig(
                seat: seats[i],
                type: PlayerType.human,
                name: 'Player ${i + 1}',
              ),
          ];

    final players = <Player>[];
    final tokens = <Token>[];

    for (var i = 0; i < slots.length; i++) {
      final slot = slots[i];
      final playerId = 'p${slot.seat.name}';
      players.add(
        Player(
          id: playerId,
          seat: slot.seat,
          name: slot.name,
          type: slot.type,
          aiDifficulty: slot.aiDifficulty,
        ),
      );
      for (var t = 0; t < 4; t++) {
        tokens.add(
          Token(
            id: '${slot.seat.name}_$t',
            seat: slot.seat,
            index: t,
          ),
        );
      }
    }

    final stamp = now ?? DateTime.now().toUtc();
    return GameState(
      id: id,
      config: config.copyWith(slots: slots, playerCount: players.length),
      players: players,
      tokens: tokens,
      startedAt: stamp,
      updatedAt: stamp,
    );
  }
}

File: src/engine/gameStateFactory.ts
import type { GameConfig, GameState, Player, Token } from './types';

export function createGameState(
  config: GameConfig,
  id = `game_${Date.now()}`,
): GameState {
  const slots =
    config.slots.length > 0
      ? config.slots
      : (['red', 'blue', 'green', 'yellow'] as const)
          .slice(0, config.playerCount)
          .map((seat, i) => ({
            seat,
            name: `Player ${i + 1}`,
            type: 'human' as const,
            aiDifficulty: 'medium' as const,
          }));

  const players: Player[] = slots.map((slot) => ({
    id: `p${slot.seat}`,
    seat: slot.seat,
    name: slot.name,
    type: slot.type,
    aiDifficulty: slot.aiDifficulty,
  }));

  const tokens: Token[] = [];
  for (const slot of slots) {
    for (let t = 0; t < 4; t++) {
      tokens.push({
        id: `${slot.seat}_${t}`,
        seat: slot.seat,
        index: t,
        status: 'yard',
        position: -1,
        pathProgress: 0,
        homeProgress: 0,
      });
    }
  }

  const now = new Date().toISOString();
  return {
    id,
    config: { ...config, slots, playerCount: players.length },
    players,
    tokens,
    phase: 'rolling',
    currentPlayerIndex: 0,
    lastDiceValue: null,
    consecutiveSixes: 0,
    winnerSeat: null,
    startedAt: now,
    updatedAt: now,
  };
}
